ciri/consensus/snapshot/local_snapshots/pruning_service.c

```diff
@@ -0,0 +1,333 @@
+
+static retcode_t collect_unconfirmed_future_transactions_for_pruning_do_func(flex_trit_t *hash, iota_stor_pack_t *pack,
+                                                                             void *data, bool *should_branch,
+                                                                             bool *should_stop) {
+  retcode_t ret = RC_OK;
+  *should_stop = false;
+  *should_branch = true;
+  collect_transactions_for_pruning_do_func_params_t *params = data;
+
+  if (pack->num_loaded == 0) {
+    *should_branch = false;
+    return RC_OK;
+  }
+
+  transaction_snapshot_index((iota_transaction_t *)pack->models[0]);
+
+  if (transaction_snapshot_index((iota_transaction_t *)pack->models[0]) == 0) {
+    if ((ret = hash243_set_add(params->transactions_to_prune, hash)) != RC_OK) {
+      return ret;
+    }
+    *should_branch = true;
+  }
+
+  return RC_OK;
+}
+
+static retcode_t collect_transactions_for_pruning_do_func(flex_trit_t *hash, iota_stor_pack_t *pack, void *data,
+                                                          bool *should_branch, bool *should_stop) {
+  retcode_t ret = RC_OK;
+  *should_stop = false;
+  *should_branch = false;
+  collect_transactions_for_pruning_do_func_params_t *params = data;
+
+  if (pack->num_loaded == 0) {
+    return RC_OK;
+  }
+
+  uint64_t current_snapshot_index = transaction_snapshot_index((iota_transaction_t *)pack->models[0]);
+
+  if (current_snapshot_index >= params->current_snapshot_index) {
+    *should_branch = true;
+    if ((ret = hash243_set_add(params->transactions_to_prune, hash)) != RC_OK) {
+      return ret;
+    }
+
+    ERR_BIND_RETURN(
+        tangle_traversal_dfs_to_future(params->tangle, collect_unconfirmed_future_transactions_for_pruning_do_func,
+                                       hash, NULL, &params),
+        ret);
+  }
+
+  if (ret) {
+    *should_stop = true;
+    return ret;
+  }
+
+  return RC_OK;
+}
+
+static void *pruning_service_routine(void *arg) {
+  pruning_service_t *ps = (pruning_service_t *)arg;
+  tangle_t tangle;
+  uint64_t start_timestamp, end_timestamp;
+  uint64_t start_index;
+
+  {
+    connection_config_t db_conf = {.db_path = ps->conf->tangle_db_path};
+
+    if (iota_tangle_init(&tangle, &db_conf) != RC_OK) {
+      log_critical(logger_id, "Failed in initializing db\n");
+      return NULL;
+    }
+  }
+
+  spent_addresses_provider_t sap;
+  {
+    connection_config_t db_conf = {.db_path = ps->spent_addresses_service->conf->spent_addresses_db_path};
+
+    if (ps->spent_addresses_service->conf->spent_addresses_files == NULL) {
+      return RC_OK;
+    }
+
+    if (iota_spent_addresses_provider_init(&sap, &db_conf) != RC_OK) {
+      log_error(logger_id, "Initializing spent addresses database connection failed\n");
+      goto cleanup;
+    }
+  }
+
+  DECLARE_PACK_SINGLE_MILESTONE(milestone, milestone_ptr, milestone_pack);
+  iota_tangle_milestone_load_first(&tangle, &milestone_pack);
+
+  lock_handle_lock(&ps->lock_handle);
```

Comment:
The lock doesn't really have to be in the struct, for the other components we declare it directly on the routine

Reply:
no, it's used from the `local_snapshots_manager` thread and it's protecting `last_snapshot_index_to_prune`

Reply:
You can't use it for both the data and the conditional variable

Reply:
Cond var needs to have its own